### PR TITLE
OnnxRuntime QNN EP

### PR DESCRIPTION
**Pull request by HectorSVC**
### Description
Integrate Qualcomm QNN SDK to enable inference on QC hexagon NPU devices

### Motivation and Context
Enable Ort inference on QC hexagon NPU devices.

### CODE REVIEW

File: tools/ci_build/github/azure-pipelines/android-arm64-v8a-QNN-crosscompile-ci-pipeline.yml [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  - task: UsePythonVersion@0
+    displayName: Use Python $(pythonVersion)
+    inputs:
+      versionSpec: $(pythonVersion)
+
+  - script: sudo apt-get update -y && sudo apt-get install -y coreutils ninja-build
+    displayName: Install coreutils and ninja
+
+  - script: /bin/bash tools/ci_build/github/android/setup_gradle_wrapper.sh $(pwd)
+    displayName: Setup gradle wrapper to use gradle 6.8.3
```
Comment (adrianlizarraga) [inline tools/ci_build/github/azure-pipelines/android-arm64-v8a-QNN-crosscompile-ci-pipeline.yml:18]: This file no longer exists. It was removed by https://github.com/microsoft/onnxruntime/pull/14733
We now have to use a template.

```suggestion
  - template: templates/set-up-gradle-wrapper-step.yml
```

File: onnxruntime/test/util/default_providers.cc [hunk 2]
```diff
@@ -209,6 +209,22 @@ std::unique_ptr<IExecutionProvider> DefaultSnpeExecutionProvider() {
 #endif
 }
 
+std::unique_ptr<IExecutionProvider> DefaultQnnExecutionProvider() {
+#if defined(USE_QNN)
+  ProviderOptions provider_options_map;
+  // Limit to CPU backend for now. TODO: Enable HTP emulator
+  std::string backend_path = "./libQnnCpu.so";
+#if defined(_WIN32) or defined(_WIN64)
```
Comment (adrianlizarraga) [inline onnxruntime/test/util/default_providers.cc:217]: I didn't know you could use `or` in preprocessor macros. Should probably just use `||` for consistency.

File: tools/ci_build/github/azure-pipelines/linux-qnn-ci-pipeline.yml [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      - script: |
+          env | egrep -e QNN
+          python3 tools/ci_build/build.py \
+            --build_dir build \
+            --config Release \
+            --parallel \
+            --use_qnn \
+            --qnn_home $QNN_SDK_ROOT \
+            --cmake_generator=Ninja \
+            --skip_tests \
```
Comment (adrianlizarraga) [inline tools/ci_build/github/azure-pipelines/linux-qnn-ci-pipeline.yml:35]:
```suggestion
            --skip_tests
```

File: onnxruntime/core/providers/qnn/builder/opbuilder/pool_op_builder.cc [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    scalar_param.dataType = QNN_DATATYPE_BOOL_8;
+    scalar_param.bool8Value = 1;
+    QnnParamWrapper count_pad_for_edges_param(node_unit.Index(),
+                                              node_unit.Name(), 
+                                              qnn_def::count_pad_for_edges,
+                                              scalar_param);
+    param_tensor_names.push_back(count_pad_for_edges_param.GetParamTensorName());
+    qnn_model_wrapper.AddParamWrapper(std::move(count_pad_for_edges_param));
+  }
+  this->output_count_ = 1;
```
Comment (pranavsharma) [inline onnxruntime/core/providers/qnn/builder/opbuilder/pool_op_builder.cc:205]: Using ```this->``` is redundant.

File: onnxruntime/core/providers/qnn/builder/qnn_def.h [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+#include "core/common/common.h"
+
+namespace onnxruntime {
+namespace qnn {
+// QNN only support subset of POSIX of dlopen/dlsym/dladdr/dlerror/dlclose
+// except the following flags for dlopen, others should be done only
+// when we really need them
+// DL_NOW is MUST
+// DL_LOCAL is enabled if not specified
+enum {
```
Comment (pranavsharma) [inline onnxruntime/core/providers/qnn/builder/qnn_def.h:22]: Always prefer enum class. Applies to other places as well.

**Review by pranavsharma — COMMENTED** [review]
Is there a class diagram or a sequence diagram that makes the interactions between the classes clear for both reviewers and maintainers?

File: onnxruntime/core/providers/qnn/builder/opbuilder/base_op_builder.h [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+    }
+  }
+
+  Status ProcessAxisAttribute(const QnnModelWrapper& qnn_model_wrapper,
+                              const NodeUnit& node_unit,
+                              Qnn_Scalar_t& axis_qnn_scalar,
+                              int32_t& default_axis_value) const;
+  Qnn_TensorType_t GetInputTensorType(const QnnModelWrapper& qnn_model_wrapper, const std::string& input_name) const;
+
+  mutable size_t output_count_ = std::numeric_limits<uint32_t>::max();
```
[the remark this answers is not in the document]
Comment (HectorSVC) [reply onnxruntime/core/providers/qnn/builder/opbuilder/base_op_builder.h:251]: need to change it in a pool_op_builder

**Comment by HectorSVC** [comment]
> Is there a class diagram or a sequence diagram that makes the interactions between the classes clear for both reviewers and maintainers?

We have a design doc. Will update it and put it somewhere in future when we come back to the docs work.

**Comment by jywu-msft** [comment]
Do the pipelines still need to be created? I see the yml files but no CI's associated with them running as part of this PR?

**Comment by HectorSVC** [comment]
> Do the pipelines still need to be created? I see the yml files but no CI's associated with them running as part of this PR?

We have new pipeline created with name prefix Github. Thanks Adrian.

File: onnxruntime/core/providers/qnn/builder/opbuilder/conv_op_builder.cc [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+
+  const auto* type_proto = outputs[0].node_arg.TypeAsProto();
+  Qnn_DataType_t qnn_data_type = QNN_DATATYPE_FLOAT_32;
+  ORT_RETURN_IF_ERROR(GetQnnDataType(is_quantized_model, type_proto, qnn_data_type));
+  ORT_RETURN_IF_NOT(qnn_model_wrapper.ProcessQuantizationParameter(outputs[0].quant_param,
+                                                                   output_quantize_param.scaleOffsetEncoding.scale,
+                                                                   output_quantize_param.scaleOffsetEncoding.offset),
+                    "Cannot get quantization parameter");
+
+  const auto group = node_helper.Get("group", (int32_t)1);
```
Comment (adrianlizarraga) [inline onnxruntime/core/providers/qnn/builder/opbuilder/conv_op_builder.cc:340]:
```suggestion
  const uint32_t group = SafeInt<uint32_t>(node_helper.Get("group", static_cast<int64_t>(1)));
```

File: onnxruntime/core/providers/qnn/builder/opbuilder/conv_op_builder.cc [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+  uint32_t num_output = output_shape[3];
+  uint32_t num_input_channel = 0;
+  ORT_RETURN_IF_ERROR(GetInputChannelNumber(qnn_model_wrapper, node_unit, num_input_channel));
+  LOGS(logger, VERBOSE) << (node_unit.OpType() == "Conv" ? "Conv:" : "ConvTranspose:")
+                        << " num_output: " << num_output << ", num_input_channel: " << num_input_channel << ", group: " << group;
+  const static std::string depthwise_conv2d = "DepthWiseConv2d";
+  bool is_depthwise_conv2d = false;
+  if ((node_unit.OpType() == "Conv") && (num_input_channel == num_output) && (static_cast<const uint32_t>(group) == num_output)) {
```
Comment (adrianlizarraga) [inline onnxruntime/core/providers/qnn/builder/opbuilder/conv_op_builder.cc:348]:
```suggestion
  if ((node_unit.OpType() == "Conv") && (num_input_channel == num_output) && (group == num_output)) {
```

File: onnxruntime/core/providers/qnn/builder/opbuilder/conv_op_builder.cc [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+    is_depthwise_conv2d = true;
+  } else {  // DepthWiseConv2d does not need group
+    // Conv attribute group
+    Qnn_Scalar_t group_qnn_scalar = QNN_SCALAR_INIT;
+    group_qnn_scalar.dataType = QNN_DATATYPE_UINT_32;
+    group_qnn_scalar.uint32Value = SafeInt<uint32_t>(group);
```
Comment (adrianlizarraga) [inline onnxruntime/core/providers/qnn/builder/opbuilder/conv_op_builder.cc:354]:
```suggestion
    group_qnn_scalar.uint32Value = group;
```

**Review by adrianlizarraga — COMMENTED** [review]
There's a casting warning that is now treated as an error:

```
/mnt/vss/_work/1/s/onnxruntime/core/providers/qnn/builder/opbuilder/conv_op_builder.cc:348:112: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
  348 |   if ((node_unit.OpType() == "Conv") && (num_input_channel == num_output) && (static_cast<const uint32_t>(group) == num_output)) {
      |                                                                                                                ^

```

I haven't tested these code suggestions (compiling at the moment). I'll let you know.

**Review by jywu-msft — APPROVED** [review]

**Pull request merged**